Вэбкорная версия статьи [здесь](Замыкания)

По сути, замыкание -- это функция внутри другой функции. Когда функция написана это просто текст программы. Когда она вызывается, то она становится кодом функции ПЛЮС доступные ей данные! Если обернуть функцию другой функцией, то при вызове внутренней будет доступен только контекст внешней функции, которая ограничена `{}`.

### Простой пример замыкания -- счетчик
```js
function counter() {
	let i = 0;
	return function() {
		return i++;
	}
}

const newCounter = counter();

newCounter(); // 1
newCounter(); // 2
newCounter(); // 3

const newCounter2 = counter();

newCounter2(); // 1
newCounter2(); // 2
newCounter2(); // 3
```
Значения счетчика для `newCounter1` и `newCounter2` независимы, так как создавались разными вызовами функции `counter()`, которая замкнула в себе два разных независимых контекста.

### Еще пример:

```js ln=true
function createCalcFunction(n) {
	return function() {
		console.log(1000 * n);
	}
}

const calc = createCalcFunction(42);
calc();
```

Создаем функцию `createCalcFunction`, которая возвращает другую функцию, использующую переданный аргумент.

На строке `7` создаем переменную в которую записываем результат отработки первой функции с аргументом `42`.

Когда мы вызываем созданную через переменную функцию `calc()` без аргументов, то она вернёт в консоль `42000`, так как запомнила, что в неё передавали значение `42`, которое она умножила на `1000`, то есть, замкнула в себя его.

### Другой пример

```js ln=true
function createIncrementor(n) {
	return function(num) {
		return n + num;
	}
}

const addOne = createIncrementor(1);
console.log(addOne(10)); // 11
console.log(addOne(41)); // 42

const addTen = createIncrementor(10);
console.log(addTen(10)); // 20
console.log(addTen(41)); // 52
```

При создании `addOne` функция `createIncrementor` замкнула в себе переданную единицу. Естественно, при вызове `addOne` результат будет увеличен на единицу. А при создании `addTen` замкнули десятку и она будет возвращать увеличение на десять.

### Еще пример
```js ln=true
function urlGenerator(domain) {
	return function(url) {
		return `https://${url}.${domain}`;
	}
}

const comUrl = urlGenerator('.com');
const ruUrl = urlGenerator('.ru');

console.log(comUrl(google)); // https://google.com
console.log(comUrl(netflix)); // https://netflix.com
console.log(ruUrl(vk)); // https://vk.ru
console.log(ruUrl(yandex)); // https://yandex.ru
```
Функция добавит указанный урл в начало домена.

### Еще пример. Немного велосипедный, но всё же
```js ln=true
const person1 = {name: "Roman", age: 35, job: "Frontend"};
const person2 = {name: "Natusya", age: 30, job: "Pharm"};

function logPerson() {
    console.log(`Person: ${this.name}, ${this.age}, ${this.job}`);
}

function bind(context, func) {
  return function(...args) {
    func.apply(context, args);
  }
}

bind(person1, logPerson)();
bind(person2, logPerson)();
```

Создадим свой велосипедный аналог метода `.bind()` на основе метода `.apply()`. Нам нужно, чтобы наша функция `bind()` привязывала контекст переданного в неё аргумента и вызывала переданную функцию с указанным контекстом.

Строки `8-12`: в функцию передается, собственно, контекст и функция. И возвращается функция, замыкающая переданный контекст. В неё указываются параметры через спрэд (`...`), чтобы можно было передать неограниченное количество параметров в виде массива для метода `.apply()` вторым аргументом. Что, собственно, и происходит на 10 строке. К переданной на строке 8 функции func апплаится переданный там же контексти непонятно зачем массив с аргументами... Ну да бог с ними. Можно и без них, тоже будет работать.

### Интересный пример функции, вызываемой только первый раз, сколько бы впоследствии её не вызывали бы.

```js
const once = fn => {
  let isExecuted = false;

  return function() {
    if (!isExecuted) {
      isExecuted = true;
      return fn();
    }
  }
};

const f = () => console.log('hi!');
const onceF = once(f);
onceF(); // hi!
onceF(); // nothing
```
